posts/2024-05-06-check-cfg.md

```diff
@@ -0,0 +1,124 @@
+
+## A look at the feature 
+
+Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with its [well known cfgs](https://doc.rust-lang.org/nightly/rustc/check-cfg.html#well-known-names-and-values) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.
+
+*`Cargo.toml`*:
+    
+```toml
+[package]
+name = "foo"
+
+[features]
+lasers = []
+zapping = []
+```
+
+*`src/lib.rs`:*
+
+```rust
+#[cfg(feature = "lasers")]  // This condition is expected
+                            // as "lasers" is an expected value
+                            // of the `feature` cfg
+fn shoot_lasers() {}
+
+#[cfg(feature = "monkeys")] // This condition is UNEXPECTED
+                            // as "monkeys" is NOT an expected
+                            // value of the `feature` cfg
+fn write_shakespeare() {}
+
+#[cfg(windosw)]             // This condition is UNEXPECTED
+                            // it's supposed to be `windows`
+fn win() {}
+```
+
+*`cargo check`*:
+
+![cargo-check](../../../../images/2024-05-06-check-cfg/cargo-check.svg)
+
+## Custom cfgs and build scripts
+
+> In Cargo point-of-view: a custom cfg is one that is neither defined by `rustc` nor by a Cargo feature. Think of `tokio_unstable`, `has_foo`, ... but not `feature = "lasers"`, `unix` or `debug_assertions`
+
+Some crates use custom cfgs that they either expected from the environment (`RUSTFLAGS`or other means) or is enabled by some logic in the crate `build.rs`. For those crates Cargo provides a new instruction: [`cargo::rustc-check-cfg`](https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#rustc-check-cfg)[^2] (or `cargo:rustc-check-cfg` for older Cargo version).
+
+[^2]: `cargo::rustc-check-cfg` will start working in Rust 1.80 (or nightly-2024-05-05). From Rust 1.77 to Rust 1.79 *(inclusive)* it is silently ignored. In Rust 1.76 and below a warning is emitted when used without the unstable Cargo flag `-Zcheck-cfg`.
+
+The syntax to use is described in the [rustc book](https://doc.rust-lang.org/nightly/rustc/) section [checking configuration](https://doc.rust-lang.org/nightly/rustc/check-cfg.html), but in a nutshell the basic syntax of `--check-cfg` is:
+
+```
+cfg(name, values("value1", "value2", ..., "valueN"))
+```
+
+Note that every custom cfgs must always be expected, regardless if the cfg is active or not!
+
+### `build.rs` example
+
+`build.rs`:
+```rust
+fn main() {
+    println!("cargo::rustc-check-cfg=cfg(has_foo)");
+    //        ^^^^^^^^^^^^^^^^^^^^^^ new with Cargo 1.80
+    if has_foo() {
+        println!("cargo::rustc-cfg=has_foo");
+    }
+}
+```
+
+> Each `cargo::rustc-cfg` must always have a accompanying **unconditional** `cargo::rustc-check-cfg` directive otherwise you risk having warnings like this: `unexpected cfg condition name: has_foo`.
```

Comment:
```suggestion
> Each `cargo::rustc-cfg` should have an accompanying **unconditional** `cargo::rustc-check-cfg` directive to avoid warnings like this: `unexpected cfg condition name: has_foo`.
```
- grammar
- The "must" language makes this sound like its generating errors. Yes, it later says warning but setting expectations early in the statement would help